File: backjoon_level_c/backjoon_level_c/1463.c
#include <stdio.h>



int dp[1000001];

int minRecur(int n);

int main() {
	int input;


	scanf("%d", &input);

	for (int i = 1; i <= input; i++) {
		dp[i] = -1;
	}
	printf("%d", minRecur(input));

	return 0;
	 

}

int minRecur(int n) {
	if (n == 1) {
		return 0;
	}
	
	if (dp[n] == -1) {
		if (n % 3 == 0 && n % 2 == 0) {
			dp[n] = min3(minRecur(n - 1), minRecur(n / 3), minRecur(n / 2));
			dp[n] = dp[n] + 1;
		}
		else if (n % 3 == 0 && n % 2 != 0) {
			dp[n] = min2(minRecur(n - 1), minRecur(n / 3));
			dp[n] = dp[n] + 1;
		}
		else if (n % 3 != 0 && n % 2 == 0) {
			dp[n] = min2(minRecur(n - 1), minRecur(n / 2));
			dp[n] = dp[n] + 1;
		}
		else {
			dp[n] = minRecur(n - 1);
			dp[n] = dp[n] + 1;
		}
		
	}
	return dp[n];
	
	
	

}

int min3(int a, int b, int c) {
	
	int min = min2(a, b);
	if (c < min) {
		min = c;
	}
	return min;
}

int min2(int a, int b) {
	if (a > b)
		return b;
	else
		return a;
}

File: backjoon_level_python/1463.py
import sys
sys.setrecursionlimit(10**7)


def min_recur(num, dp):
    m = sys.maxsize

    if dp[num] != -1 or num == 1:
        return dp[num]

    if num % 3 == 0:
        m = min(m, min_recur(num//3,dp))
    if num % 2 == 0:
        m = min(m, min_recur(num//2,dp))
    m = min(m, min_recur(num - 1, dp))
    dp[num] = m+1
    return dp[num]


def solve(num):
    dp = [-1] * 1000001
    dp[1] = 0
    min_recur(num,dp)
    return dp[num]


def solve2(num):
    dp = [0] * 1000001
    dp[1] = 0
    dp[2] = 1
    dp[3] = 1

    for i in range(4, 1000001):
        c = sys.maxsize
        if i%3 == 0:
            c = min(c, dp[i//3] + 1)
        if i%2 == 0:
            c = min(c, dp[i//2] + 1)
        c = min(c, dp[i-1]+1)
        dp[i] = c
    return dp[num]


if __name__ == '__main__':
    num = int(input())

    print(solve(num))